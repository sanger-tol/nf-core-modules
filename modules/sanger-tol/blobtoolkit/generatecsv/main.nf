process BLOBTOOLKIT_GENERATECSV {
    tag "${[meta, meta2, meta3].collect { it.id }.findAll().unique()[0]}"

    input:
    // It is expected that each meta will contain
    // meta.library_layout : {true/false}
    // meta.datatype: {pacbio/ont/illumina}
    tuple val(meta),    path(pacbio_files)
    tuple val(meta2),   path(ont_files)
    tuple val(meta3),   path(illumina_files)

    output:
    tuple val(new_meta),    path("${prefix}.samplesheet.csv")   , emit: csv
    path("versions.yml")                                    , emit: versions

    when:
    task.ext.when == null || task.ext.when

    exec:
    // Note: Manually bump version number when updating module
    def VERSION       = "1.0.0"

    // CHECK META.ID, IF 1 NOT SET THEN TRY NEXT
    def sample_id = meta.id ?: meta2.id ?: meta3.id
    prefix = task.ext.prefix ?: sample_id

    // SET NEW META OBJECT
    new_meta = meta ?: meta2 ?: meta3

    // CHECK ALL FOR DATA
    def pacbio_data   = pacbio_files.findAll { it.name.endsWith('.fasta.gz') || it.name.endsWith('.fa.gz') }
    def ont_data      = ont_files.findAll { it.name.endsWith('.fasta.gz') || it.name.endsWith('.fa.gz') }
    def illumina_data = illumina_files.findAll { it.name.endsWith('.fasta.gz') || it.name.endsWith('.fa.gz') }

    // ENSURE WE HAVE AT LEAST 1 FILE
    assert (pacbio_data.size() + ont_data.size() + illumina_data.size()) > 0

    // SET HEADER
    def samplesheet_entries = ["sample,datatype,datafile,library_layout"]

    // START TOTAL FILE COUNTER
    def counter = 0

    // FORCE PREFIX FOR ALL THE BELOW TO ENSURE SAME NAME
    pacbio_data.eachWithIndex { fa, idx ->
        samplesheet_entries << "${prefix}_T${idx+1+counter},${meta.data_type},${fa},${meta.library_layout}"
    }
    counter += pacbio_data.size()

    ont_data.eachWithIndex { fa, idx ->
        samplesheet_entries << "${prefix}_T${idx+1+counter},${meta2.data_type},${fa},${meta2.library_layout}"
    }
    counter += ont_data.size()

    illumina_data.eachWithIndex { fa, idx ->
        samplesheet_entries << "${prefix}_T${idx+1+counter},${meta3.data_type},${fa},${meta3.library_layout}"
    }

    file(task.workDir.resolve("${prefix}.samplesheet.csv")).text = samplesheet_entries.join("\n")

    file("${task.workDir}/versions.yml").text = """\
        BLOBTOOLKIT_GENERATECSV:
            blobtoolkit_generatecsv: ${VERSION}
        """.stripIndent()
}
