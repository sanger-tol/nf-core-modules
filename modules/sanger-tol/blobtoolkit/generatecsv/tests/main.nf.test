nextflow_process {

    name "Test Process BLOBTOOLKIT_GENERATECSV"
    script "../main.nf"
    process "BLOBTOOLKIT_GENERATECSV"

    tag "modules"
    tag "modules_sangertol"
    tag "sangertol"
    tag "blobtoolkit/generatecsv"

    test("pacbio fasta.gz to csv") {

        when {
            process {
                """
                input[0] = [
                    [ id:'test', data_type:'pacbio', library_layout: 'SINGLE' ], // meta map
                    [
                        file(params.modules_testdata_base_path + 'resources/ascc/asccTinyTest_V2/pacbio/Pyoeliiyoelii17XNL_pbtools_simulated_pacbio_reads.fa.gz', checkIfExists: true),
                        file(params.modules_testdata_base_path + 'resources/treeval/TreeValTinyData/genomic_data/pacbio/seqkitPacbio50000.fasta.gz', checkIfExists: true)
                    ]
                ]
                input[1] = [[],[]]
                input[2] = [[],[]]
                """
            }
        }

        then {
            assertAll(
                { assert process.success },
                { assert snapshot(process.out).match() }
            )
        }
    }

    test("ont fasta.gz to csv") {

        when {
            process {
                """
                input[0] = [[],[]]
                input[1] = [
                    [ id:'test', data_type:'ont', library_layout: 'SINGLE' ], // meta map
                    [ file(params.modules_testdata_base_path + 'resources/ascc/asccTinyTest_V2/pacbio/Pyoeliiyoelii17XNL_pbtools_simulated_pacbio_reads.fa.gz', checkIfExists: true) ]
                ]
                input[2] = [[],[]]
                """
            }
        }

        then {
            assertAll(
                { assert process.success },
                { assert snapshot(process.out).match() }
            )
        }

    }

    test("multiple fasta.gz to csv") {

        when {
            process {
                """
                input[0] = [[],[]]
                input[1] = [
                    [ id:'test', data_type:'ont', library_layout: 'SINGLE' ], // meta map
                    [ file(params.modules_testdata_base_path + 'resources/ascc/asccTinyTest_V2/pacbio/Pyoeliiyoelii17XNL_pbtools_simulated_pacbio_reads.fa.gz', checkIfExists: true) ]
                ]
                input[2] = [
                    [ id:'test', data_type:'illumina', library_layout: 'SINGLE' ], // meta map
                    [ file(params.modules_testdata_base_path + 'resources/treeval/TreeValTinyData/genomic_data/pacbio/seqkitPacbio50000.fasta.gz', checkIfExists: true) ]
                ]
                """
            }
        }

        then {
            assertAll(
                { assert process.success },
                { assert snapshot(process.out).match() }
            )
        }

    }

}
