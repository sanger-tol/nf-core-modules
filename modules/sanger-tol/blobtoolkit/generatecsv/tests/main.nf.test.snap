{
    "ont fasta.gz to csv": {
        "content": [
            {
                "0": [
                    [
                        {
                            "id": "test",
                            "data_type": "ont",
                            "library_layout": "SINGLE"
                        },
                        "test.samplesheet.csv:md5,b07de6dacdd5c4f314047496e3d32a1b"
                    ]
                ],
                "1": [
                    "versions.yml:md5,f7b11b15294a6105291c8c0be9118efb"
                ],
                "csv": [
                    [
                        {
                            "id": "test",
                            "data_type": "ont",
                            "library_layout": "SINGLE"
                        },
                        "test.samplesheet.csv:md5,b07de6dacdd5c4f314047496e3d32a1b"
                    ]
                ],
                "versions": [
                    "versions.yml:md5,f7b11b15294a6105291c8c0be9118efb"
                ]
            }
        ],
        "meta": {
            "nf-test": "0.9.2",
            "nextflow": "25.04.6"
        },
        "timestamp": "2025-10-31T15:08:56.481170419"
    },
    "multiple fasta.gz to csv": {
        "content": [
            {
                "0": [
                    [
                        {
                            "id": "test",
                            "data_type": "ont",
                            "library_layout": "SINGLE"
                        },
                        "test.samplesheet.csv:md5,c36e7caa51f227defe1dde5aa1cc1e60"
                    ]
                ],
                "1": [
                    "versions.yml:md5,f7b11b15294a6105291c8c0be9118efb"
                ],
                "csv": [
                    [
                        {
                            "id": "test",
                            "data_type": "ont",
                            "library_layout": "SINGLE"
                        },
                        "test.samplesheet.csv:md5,c36e7caa51f227defe1dde5aa1cc1e60"
                    ]
                ],
                "versions": [
                    "versions.yml:md5,f7b11b15294a6105291c8c0be9118efb"
                ]
            }
        ],
        "meta": {
            "nf-test": "0.9.2",
            "nextflow": "25.04.6"
        },
        "timestamp": "2025-10-31T15:11:54.091240254"
    },
    "pacbio fasta.gz to csv": {
        "content": [
            {
                "0": [
                    [
                        {
                            "id": "test",
                            "data_type": "pacbio",
                            "library_layout": "SINGLE"
                        },
                        "test.samplesheet.csv:md5,85f28ee33669df8cbe5c945988879e54"
                    ]
                ],
                "1": [
                    "versions.yml:md5,f7b11b15294a6105291c8c0be9118efb"
                ],
                "csv": [
                    [
                        {
                            "id": "test",
                            "data_type": "pacbio",
                            "library_layout": "SINGLE"
                        },
                        "test.samplesheet.csv:md5,85f28ee33669df8cbe5c945988879e54"
                    ]
                ],
                "versions": [
                    "versions.yml:md5,f7b11b15294a6105291c8c0be9118efb"
                ]
            }
        ],
        "meta": {
            "nf-test": "0.9.2",
            "nextflow": "25.04.6"
        },
        "timestamp": "2025-10-31T15:08:47.825237422"
    },
    "fasta.gz to csv": {
        "content": [
            {
                "0": [
                    [
                        {
                            "id": "test"
                        },
                        "test.samplesheet.csv:md5,85f28ee33669df8cbe5c945988879e54"
                    ]
                ],
                "1": [
                    "versions.yml:md5,d4f404227c9dfc0dfd7d27fbf1e33557"
                ],
                "csv": [
                    [
                        {
                            "id": "test"
                        },
                        "test.samplesheet.csv:md5,85f28ee33669df8cbe5c945988879e54"
                    ]
                ],
                "versions": [
                    "versions.yml:md5,d4f404227c9dfc0dfd7d27fbf1e33557"
                ]
            }
        ],
        "meta": {
            "nf-test": "0.9.2",
            "nextflow": "25.04.6"
        },
        "timestamp": "2025-10-30T16:53:42.356305559"
    }
}